src/driver/xdma_sgdma.c

```diff
@@ -261,22 +263,51 @@ ssize_t xdma_channel_read_write(struct xdma_channel *chnl,
 		return -EINVAL;
 	}
 
-	rv = check_transfer_align(engine, buf, count, *pos, 1);
-	if (rv) {
-		pr_info("Invalid transfer alignment detected\n");
-		return rv;
-	}
+	if (sg_alloc_table(&sgt, SG_MAX_SINGLE_ALLOC, GFP_KERNEL))
+		return -ENOMEM;
+
+	while (max_bytes && iov_iter_count(iter)) {
+		pages = (struct page **) sgt.sgl;
+		pages += SG_MAX_SINGLE_ALLOC * (PAGE_PTRS_PER_SGL - 1);
+
+		size = iov_iter_get_pages(iter, pages, max_bytes, SG_MAX_SINGLE_ALLOC, &offset);
+		if (size < 0)
+			return size;
+
+		for (left = size, i = 0; left > 0; left -= len, i++) {
+			len = min_t(size_t, PAGE_SIZE - offset, left);
+			sg_set_page(&sgt.sgl[i], pages[i], len, offset);
+			offset = 0;
+		}
+		pages_nr = i;
 
-	memset(&cb, 0, sizeof(struct xdma_io_cb));
-	cb.buf = (char __user *)buf;
-	cb.len = count;
-	rv = xdma_channel_map_user_buf_to_sgl(&cb, write);
-	if (rv < 0)
-		return rv;
+		rv = check_transfer_align(engine, (void *) offset, size, pos, 1);
+
+		if (rv) {
+			pr_info("Invalid transfer alignment detected\n");
+			return rv;
+		}
+
+		memset(&sgt.sgl[i], 0, (SG_MAX_SINGLE_ALLOC - pages_nr) *
```

Comment:
This memset is highly suspect. I think you want to use sg_mark_end()...

Reply:
Yup. Plus there was a bug on the original xdma code that was also needed

Reply:
that needed to be fixed*